### PR TITLE
feat(renderers): C-108.x — Renderer interface + DashboardRenderer + PagerDutyRenderer (MIG-7.2d)

### PR DESCRIPTION
**Pull request by jcfischer**
## Summary

Adds the non-agent-bound surface layer per architecture §9.2: renderers subscribe to bus subjects and project envelopes into sinks (UI, pager, log stream). Distinct from presence adapters — renderers are deployment-scoped activity-centric projections.

The G-1111 §4.6 fail-safe rule requires ≥2 distinct platform classes covering `local.{org}.system.>` so an operational alert reliably reaches the operator even if one sink is the thing that broke. This PR ships the "recommended pair" (dashboard + pagerduty); cli-tail and webhook-out are scheduled for follow-on once a real cortex deployment surfaces the need.

## Migration context

cortex#9 / `docs/plan-cortex-migration.md` §4 MIG-7.2d. After this PR the platform half of MIG-7.2* is complete (#45 binding, #46 internal, #47 flip, #48 cleanup, #49 mattermost). Lifting surface fields out of `DiscordAdapterInfra` / `MattermostAdapterInfra` into dedicated `kind: discord-channel` / `mattermost-channel` renderer kinds is a follow-on iteration that depends on schema decisions for those new kinds.

## Refactor scope

**`src/renderers/` (NEW, +395 LOC + 270 LOC tests):**
- `types.ts` — `Renderer` interface: extends SurfaceAdapter contract with lifecycle. Documents idempotent start, render-must-not-throw, no bus side-effects.
- `pagerduty.ts` — real PagerDuty events-v2 client. Mapping: event_action=trigger, deterministic dedup_key, severity classifier (crashed→critical, degraded→error, fallback→warning) with `payload.severity` override, summary truncated at PagerDuty's 1024-char hard limit, full envelope in custom_details. Failure handling: HTTP non-2xx + thrown errors log + drop. AbortSignal composition combines router signal with own 5s timeout. Configurable `endpoint` + `fetchImpl` for tests.
- `dashboard.ts` — in-memory ring buffer (default 1000 entries). Tests probe `getRecent()` to assert delivery; Mission Control integration deferred to MIG-7.13 (existing `src/surface/mc/` reads from the buffer once MC wires into top-level startup). Bounded-write semantics.
- `index.ts` — `createRenderer(config)` factory dispatching on `kind`. `UnimplementedRendererKindError` for cli-tail/webhook-out with kind-named message. Exhaustive `default` branch with TS `never` check.

**24 tests (3 files):**
- `pagerduty.test.ts` — 11 tests (POST URL, payload shape, routing_key, event_action, dedup_key, severity classification, summary truncation, custom_details, failure paths, surfaceConfig, custom endpoint)
- `dashboard.test.ts` — 7 tests (insertion order, bounded buffer + drop-oldest, snapshot isolation, stop drains, malformed envelope tolerance, surfaceConfig, idempotent start)
- `factory.test.ts` — 5 tests (dashboard/pagerduty dispatch, cli-tail/webhook-out throw, error message)

**`src/cortex.ts` (+47):**
- new loop after platform-adapter blocks: for each renderer config, strict parse via `RendererSchema.safeParse`, build via factory, `start()`, register surfaceConfig with router. Log + skip individual failures per the G-1111 fail-safe rule.
- shutdown integration: renderer-stop slots register alongside adapter-stop slots in the named-shutdown drain.

**Schema additions:**
- `src/common/types/config.ts` — additive optional `renderers: z.array(z.unknown()).optional()` on BotConfig so existing bot.yaml loads unchanged; strict per-variant parse happens at runtime in cortex.ts before the factory call.
- `src/common/types/cortex-config.ts` — renamed `Renderer` → `RendererConfig` (the schema type represents the config block, not the renderer instance; unprefixed `Renderer` now refers to the runtime class hierarchy). Single greppable site; no in-tree consumers besides this PR.

## Test plan

- [x] `bunx tsc --noEmit` clean
- [x] `bun test src/renderers/` — 24/24 pass
- [x] `bun test` (full suite) — 2006 pass, 1 pre-existing mission-control React-shell fail (unchanged; gated on MIG-7 cutover)
- [x] shutdown drain integration: renderer slots register in the named shutdown sequence and stop cleanly within 15s default timeout

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/renderers/dashboard.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    this.buffer.push(envelope);
+    if (this.buffer.length > this.bufferSize) {
+      // Drop the oldest envelope to maintain the bound. Single shift per
+      // overflow is O(n) on the underlying array; if the buffer hot path
+      // ever becomes a profile concern we'd switch to a true ring index
+      // here, but at 1000-entry default + single-digit envelope rate this
+      // is several orders of magnitude below any measurable cost.
+      this.buffer.shift();
+    }
+  }
```
Comment (jcfischer) [inline src/renderers/dashboard.ts:106]: **[warning/quality] render() lacks try-catch despite "never throw" contract**

PagerDutyRenderer wraps `render()` in try-catch to honor Renderer contract §2. DashboardRenderer doesn't — it relies on `push()` and `shift()` being infallible. That's true today, but the contract is *structural*. If someone adds envelope field access here later (e.g., logging `envelope.id` for debug), a malformed envelope breaks the "never throw" guarantee silently.

The test at dashboard.test.ts:70 even exercises null input, proving the path matters.

Fix: wrap the body in try-catch like PagerDutyRenderer does — one catch, one `console.warn`, zero cost for the happy path.

File: src/renderers/dashboard.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      subjects: this.subjects,
+      render: (envelope, signal) => this.render(envelope, signal),
+    };
+  }
+
+  async render(envelope: Envelope, _signal?: AbortSignal): Promise<void> {
+    // Renderer contract §2: never throw. The ring-buffer push is pure
+    // synchronous work; the only failure path is OOM under absurd
+    // bufferSize values, which is a config bug, not a runtime concern.
+    this.buffer.push(envelope);
```
Comment (jcfischer) [inline src/renderers/dashboard.ts]: **[warning/quality] render() silently buffers null/malformed envelopes**

The null-envelope test (dashboard.test.ts:70) proves `render()` won't throw on null — but it *does* push null into the buffer. `getRecent()` then returns it in the `readonly Envelope[]` array, and the consumer (eventually Mission Control) gets a null disguised as an Envelope.

The surface-router validates before dispatching, so this is defense-in-depth — but the contract says "never throw," not "silently buffer garbage."

Fix: `if (!envelope) return;` guard before the push.

File: src/cortex.ts [hunk 3]
```diff
@@ -329,6 +331,46 @@ export async function startCortex(
...
+    } catch (err) {
+      // Renderer construction can throw (UnimplementedRendererKindError)
+      // or `start()` can fail. Per the G-1111 fail-safe rule, ONE broken
+      // renderer must not prevent the others from coming up — log + skip.
+      console.error(
+        `cortex: ${rendererConfig.kind} renderer failed to start:`,
+        err instanceof Error ? err.message : err,
+      );
+    }
+  }
```
Comment (jcfischer) [inline src/cortex.ts:372]: **[warning/architecture] Renderers not wired into ConfigWatcher hot-reload**

The config watcher at cortex.ts:392 calls `adapter.updateConfig?.(event.config)` for platform adapters, but renderers are skipped. The Renderer interface has no `updateConfig` method, so this is by design — but the consequence is that a PagerDuty routing key change or dashboard subscription change in bot.yaml requires a full cortex restart.

If restart-only is the intent (reasonable for renderers), add a one-liner in the ConfigWatcher block:
```
// Renderers are restart-only — no hot-reload (routing key / subscribe changes require bounce).
```

If hot-reload should eventually reach renderers, open a follow-on issue rather than silently shipping the gap.

**Comment by jcfischer** [comment]
## Full Review: the-metafactory/cortex#50

### Lens Results
| Lens | Critical | Warning | Suggestion | Nit | Verdict |
|------|----------|---------|------------|-----|---------|
| CodeQuality | 0 | 2 | 1 | 0 | advisory |
| Security | 0 | 0 | 0 | 0 | pass |
| Hardening | 0 | 0 | 0 | 0 | pass (not applicable — no new endpoints) |
| Architecture | 0 | 1 | 0 | 1 | advisory |
| EcosystemCompliance | 0 | 0 | 0 | 1 | pass |
| Performance | 0 | 0 | 0 | 0 | pass |
| Duplication | 0 | 0 | 1 | 0 | pass |

### Warnings (3)

**W1 — DashboardRenderer.render() lacks try-catch (dashboard.ts:78-91)**
PagerDutyRenderer wraps `render()` in try-catch honoring contract §2. DashboardRenderer doesn't — relies on push/shift being infallible. Structural asymmetry; a future envelope field access breaks the guarantee.

**W2 — DashboardRenderer.render() silently buffers null/malformed envelopes (dashboard.ts:82)**
Null passed to `render()` gets pushed into the buffer and returned by `getRecent()` as if it were a valid Envelope. The surface-router validates upstream, but defense-in-depth says guard before push.

**W3 — Renderers not wired into ConfigWatcher (cortex.ts:392)**
Platform adapters get `updateConfig` on hot-reload; renderers don't. Renderer interface has no such method, so this is by design. But the gap is undocumented — a PD routing key change silently has no effect until restart.

### Suggestions (2)

**S1 — Test: render() after stop() (dashboard.test.ts)**
Types.ts JSDoc says render calls post-stop SHOULD be no-ops. After `stop()`, buffer is `[]`, so `render()` re-fills it. No test pins whether that's intended or accidental.

**S2 — Shared makeEnvelope test helper (dashboard.test.ts:16 / pagerduty.test.ts:48)**
Both files create the same base Envelope with sovereignty block. A shared `__tests__/helpers.ts` would deduplicate the boilerplate. Minor — test helpers naturally diverge.

### Nits (2)

**N1 — CLAUDE.md not updated for src/renderers/**
The "Internal componentisation" section doesn't mention the new module. Low priority during the migration series.

**N2 — "dashboard stop" name collision in shutdown drain (cortex.ts:446 vs :452)**
Pre-existing `"dashboard stop"` slot (for `dashboardApi`) and new `"renderer dashboard stop"` slot coexist. Distinct strings, but reads confusingly at shutdown trace. Self-resolves at MIG-7.13.

### What I checked and didn't find issues with

- **Security:** No injection surfaces, no user-controlled URLs, routingKey stays in POST body (not logs), endpoint override is constructor-only (not config-reachable). PD events-v2 routing key authentication model is correct.
- **Performance:** Buffer bounded at 1000 with shift-based eviction. O(n) shift is ~microseconds at this scale. AbortSignal.timeout(5s) on PD fetch. getRecent() snapshot copy is intentional.
- **Architecture:** Renderer interface + factory + implementations follow the adapter pattern exactly. surfaceConfig getter, shutdown drain slots, startup loop all mirror existing adapter code. Renderer → RendererConfig rename is clean (zero consumers found). The z.unknown() → RendererSchema.safeParse two-stage validation is the right pattern for additive schema migration.
- **Type safety:** Exhaustive switch with `never` guard in factory. No `as any`. Single necessary cast on `envelope.payload` in PD renderer.
- **Test quality:** 24 tests cover happy path, failure paths, boundary conditions, snapshot isolation, surfaceConfig shapes, custom endpoint, idempotent start, malformed input tolerance.

**Review by jcfischer — COMMENTED** [review]
Full review (6 lenses + duplication) — solid new module, clean architecture matching existing adapter patterns. Three warnings: DashboardRenderer.render() missing try-catch (contract asymmetry with PagerDutyRenderer), null envelopes silently buffered, and ConfigWatcher hot-reload gap undocumented. None are runtime-breaking today but all represent defensive gaps.

verdict: blockers=0 majors=3 nits=4 — recommend: request-changes (posted as comment-review — bot account opened the PR; --request-changes blocked by GitHub)

**Comment by jcfischer** [comment]
## Full Review: the-metafactory/cortex#50

**Reviewer: Holly (adversarial review)**

### Lenses Applied (6 + duplication)

| Lens | Critical | Warning | Suggestion | Nit | Verdict |
|------|----------|---------|------------|-----|---------|
| CodeQuality | 0 | 0 | 0 | 1 | pass |
| Security | 0 | 0 | 0 | 0 | pass |
| Hardening | 0 | 0 | 0 | 0 | pass |
| Architecture | 0 | 0 | 0 | 0 | pass |
| EcosystemCompliance | 0 | 0 | 1 | 1 | advisory |
| Performance | 0 | 0 | 0 | 1 | pass |
| Duplication | 0 | 0 | 0 | 0 | pass |

---

### Lens Details

#### CodeQuality — pass (1 nit)

Error handling is solid throughout. PagerDuty's render-must-not-throw contract is upheld with proper try/catch + `console.warn`. Dashboard's `render()` relies on array push being inherently non-throwing — correct but leaves null/undefined envelopes silently buffered (see inline comment). Factory's `UnimplementedRendererKindError` + exhaustive `never` branch is exactly right. Tests are well-named and cover happy paths + failure paths. No dead code, no unused imports, no `any` types. The `as Record<string, unknown>` cast in `pagerduty.ts:139` is justified narrowing from `unknown`.

**Nit (quality/dashboard.ts:78):** `render()` silently buffers null/undefined envelopes. The test documents this intentionally ("render() never throws even on absurd input"), but `getRecent()` will return those nulls to downstream consumers. A one-line `if (!envelope) return;` guard would prevent `getRecent()` from returning junk. Not a blocker — the surface-router validates before dispatch — but the contract says "never throw", not "silently store garbage."

#### Security — pass (0 findings)

Checked injection (no string interpolation in HTTP calls — payload built programmatically), sensitive data exposure (routing key not logged, only sent in POST body where PagerDuty requires it), input validation (Zod safeParse at config boundary, severity override validated against explicit allowlist of `critical|error|warning|info`), SSRF (endpoint URL only configurable via code-level options, not from config — tests only), and auth (no new API endpoints). Clean.

#### Hardening — pass (0 findings)

No new API endpoints, no HTTP servers, no webhook consumers. Dashboard is a pure in-memory buffer. PagerDuty is outbound-only. CORS, rate limiting, auth, and PII concerns don't apply. Config validation at boundaries is handled by Zod.

#### Architecture — pass (0 findings)

Clean separation: `types.ts` (interface), `dashboard.ts` (buffer impl), `pagerduty.ts` (HTTP client), `index.ts` (factory + exports). Each module has one responsibility. Dependencies flow correctly: renderers → bus contracts → common types. No circular dependencies. The `Renderer` interface is not premature abstraction — two implementations now, two deferred, factory already dispatches four kinds. The `Renderer` → `RendererConfig` rename in cortex-config.ts is clean — grep confirms no other consumers of the old name.

The cortex.ts wiring follows the established adapter pattern: for-each → safeParse → try { create + start + register } catch { log + skip }. Shutdown drain slots mirror adapter slots. Pattern consistency is maintained.

#### EcosystemCompliance — advisory (1 suggestion, 1 nit)

**Suggestion (ecosystem/CLAUDE.md):** The CLAUDE.md architecture tree lists PagerDuty under `adapters/` but this PR moves PagerDuty to `renderers/`. The new `src/renderers/` directory is not listed in the tree. The architecture section should add `renderers/` and remove PagerDuty from the adapters line. The file says `adapters/ — platform-specific (Discord, Mattermost, Slack, PagerDuty)` — PagerDuty is now a renderer, not an adapter.

**Nit (ecosystem/PR title):** Title is ~98 chars, exceeding the 70-char guideline. The scope `(renderers)` + migration ref `(MIG-7.2d)` + three class names make it long. Consider: `feat(renderers): MIG-7.2d — Renderer interface + dashboard + PagerDuty`.

#### Performance — pass (1 nit)

**Nit (perf/dashboard.ts:86):** `Array.shift()` is O(n) per overflow. At the 1000-entry default with single-digit envelope rate this is negligible. The PR is self-aware — the comment explicitly says "switch to a true ring index" if profiling shows a concern. No action needed.

PagerDuty's per-render `AbortSignal.timeout(5s)` composed with the router's signal is the right pattern — bounded I/O, no resource leaks. Dashboard buffer is bounded with explicit cap. No growing caches, no unbounded operations, no blocking I/O.

#### Duplication — pass (0 findings)

The `surfaceConfig` getter is identical between `DashboardRenderer` and `PagerDutyRenderer`, but this follows the existing pattern in `DiscordAdapter`, `MattermostAdapter`, and `MockAdapter` — all five implementations build the same `{ id, subjects, render }` shape. The adapter versions also conditionally add `filter`. This is the codebase convention, not copy-paste debt — each implementation may diverge (e.g., adding filters) as renderers mature.

Checked PagerDuty's abort signal composition against `common/timeout.ts`'s `fetchWithTimeout` — they serve different purposes. `fetchWithTimeout` adds diagnostic error rewriting; PagerDuty needs signal composition with the router's abort + catch-all (render-must-not-throw). Not a re-implementation.

---

### What I checked but found nothing

- **Hidden coupling:** Renderers depend on bus/surface-router and common/types. No reverse dependency. The `RendererConfig` rename has zero in-tree consumers besides this PR (grep confirmed).
- **Reversibility:** BotConfig.renderers is additive optional (`z.array(z.unknown()).optional()`). Existing bot.yaml loads unchanged. The type rename `Renderer → RendererConfig` has no consumers to break. Fully reversible.
- **The thing that wasn't tested:** Render-after-stop behavior. PagerDuty stop() is a no-op, so render() after stop() would still POST. Dashboard stop() clears the buffer, so render() after stop() pushes to a fresh buffer. Neither throws, which satisfies the SHOULD-be-no-op contract. The surface-router stops dispatching before renderers stop, so this race is mostly theoretical.
- **Signal abort path in PagerDuty:** Not tested explicitly, but the surface-router's test suite covers abort signal behavior at the dispatch layer. PagerDuty forwards the composed signal to fetch, which handles abort natively.

File: src/renderers/dashboard.ts (hunk 2 above)
Comment (jcfischer) [inline src/renderers/dashboard.ts]: **[nit/quality]** `render()` silently buffers null/undefined envelopes — `this.buffer.push(envelope)` succeeds with any value, so `getRecent()` can return nulls to downstream consumers. The test "render() never throws even on absurd input" documents this intentionally, but the defense-in-depth guard is free:

```typescript
if (!envelope) return;
```

Surface-router validates before dispatch, so this only matters if a future caller bypasses the router. Not a blocker.

**Review by jcfischer — COMMENTED** [review]
Full review (6 lenses + duplication) — solid PR. Clean renderer abstraction, follows established adapter wiring pattern, well-tested (24 tests), correct shutdown integration. One CLAUDE.md documentation gap and three cosmetic nits — no correctness, security, or architecture issues found.

verdict: blockers=0 majors=0 nits=4 — recommend: request-changes (posted as comment-review — bot account opened the PR; --request-changes blocked by GitHub)

File: src/renderers/pagerduty.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+      });
+      if (!res.ok) {
+        console.warn(
+          `pagerduty-renderer: POST events-v2 returned HTTP ${res.status} ${res.statusText} for envelope ${envelope.id}`,
+        );
+      }
+    } catch (err) {
+      console.warn(
+        `pagerduty-renderer: failed to forward envelope ${envelope.id}:`,
+        err instanceof Error ? err.message : err,
```
Comment (jcfischer) [inline src/renderers/pagerduty.ts:127]: **[nit/quality]** Catch-block throw risk — asymmetry with DashboardRenderer.

`dashboard.ts:110` uses `envelope?.id ?? "<null>"` (safe). This line uses bare `envelope.id`. If a null envelope reaches `render()` (requires bypassing the surface-router — theoretical), `buildPayload(envelope)` throws → catch block accesses `envelope.id` → `TypeError` from inside the catch → unhandled rejection → contract §2 "never throw" violated.

Fix: `envelope?.id ?? "<unknown>"` to match the dashboard pattern.

**Comment by jcfischer** [comment]
## Full Review (Cycle 2): the-metafactory/cortex#50

**Reviewer: Holly (adversarial review) — re-review after 1b324fc**

### Cycle 1 Fix Verification

| Finding | Status | Verification |
|---------|--------|-------------|
| **W1** — DashboardRenderer.render() missing try-catch | **Fixed** | `dashboard.ts:78-106` now wraps push/shift in try-catch + console.warn. Structurally symmetric with PagerDutyRenderer. |
| **W2** — Null envelopes silently buffered | **Fixed** | `dashboard.ts:89` guards `if (envelope === null \|\| typeof envelope !== "object") return;` before push. New test at `dashboard.test.ts:213` pins the behavior with null, string, and undefined inputs — all three silently dropped, buffer clean. |
| **W3** — ConfigWatcher hot-reload gap undocumented | **Fixed** | `types.ts:46-57` JSDoc block explicitly documents "Renderers are static across hot-reload" with rationale (failure modes of partial hot-reload) and forward-looking note on `updateConfig` if restart cost proves too high. |
| **S1** — Test: render() after stop() behavior unpinned | **Fixed** | `dashboard.test.ts:224` new test "render() after stop() re-fills the buffer" pins the intentional behavior: stop() clears, subsequent render() re-fills. Comment cites Renderer contract §3. |
| **S2** — Shared makeEnvelope helper | **Deferred** | Acknowledged low-priority. Test helpers naturally diverge. |
| **N1** — CLAUDE.md not updated for src/renderers/ | **Deferred** | Acknowledged. Low priority during migration series. |
| **N2** — "dashboard stop" name in shutdown drain | **Deferred** | Acknowledged. Self-resolves at MIG-7.13. |

### Lens Results (Cycle 2)

| Lens | Critical | Warning | Suggestion | Nit | Verdict |
|------|----------|---------|------------|-----|---------|
| CodeQuality | 0 | 0 | 0 | 0 | pass |
| Security | 0 | 0 | 0 | 0 | pass |
| Hardening | 0 | 0 | 0 | 0 | pass (N/A — no new endpoints) |
| Architecture | 0 | 0 | 0 | 0 | pass |
| EcosystemCompliance | 0 | 0 | 0 | 0 | pass (3 carried deferrals acknowledged) |
| Performance | 0 | 0 | 0 | 0 | pass |
| Duplication | 0 | 0 | 0 | 0 | pass |

### What I verified in the cycle 2 diff

- **W1 correctness:** The catch block's `envelope?.id ?? "<null>"` is safe — envelope is guaranteed non-null at that point (the null guard returns before reaching the catch). Template literal won't throw.
- **W2 guard edge cases:** `typeof null === "object"` in JS, so the explicit `=== null` check fires first. `typeof undefined !== "object"` catches undefined. Arrays pass the guard (typeof array is "object") but are caught upstream by Zod envelope validation. Defense-in-depth, not a full validator — appropriate.
- **No regressions:** The new try-catch doesn't change the happy-path behavior (push + shift are synchronous, never throw in practice). The null guard adds one branch per render call — zero measurable cost.
- **Test count:** 26/26 per PR description (24 original + 2 new: null-drop assertion + post-stop render pin).
- **No new exports leaked:** index.ts export surface unchanged. Types.ts JSDoc addition is doc-only.

### Carried deferrals (not re-raised — already acknowledged)

These items were explicitly deferred in cycle 1 and are not new findings:
- S2: shared `makeEnvelope` helper across test files
- N1: CLAUDE.md architecture section doesn't list `src/renderers/`
- N2: "dashboard stop" vs "renderer dashboard stop" in shutdown drain naming

None affect correctness, security, or architecture. All are self-resolving or trivially fixable in a follow-on.

**Review by jcfischer — COMMENTED** [review]
Full review cycle 2 (6 lenses + duplication) — all three cycle 1 warnings fixed correctly. DashboardRenderer.render() now has try-catch + null guard (W1+W2), Renderer interface JSDoc documents static-across-hot-reload semantics (W3), new test pins post-stop render behavior (S1). Zero new findings. Three carried deferrals (S2, N1, N2) are acknowledged low-priority items.

verdict: blockers=0 majors=0 nits=0 — recommend: merge (posted as comment-review — bot account opened the PR; --approve blocked by GitHub)

**Comment by jcfischer** [comment]
## Full Review: the-metafactory/cortex#50 — Cycle 2

### Cycle 1 Fix Verification

| Finding | Status | Evidence |
|---------|--------|----------|
| W1 — DashboardRenderer.render() lacks try-catch | **FIXED** | `dashboard.ts:101-131` — try-catch wraps push/shift with `console.warn` in catch |
| W2 — null/non-object envelopes leak into buffer | **FIXED** | `dashboard.ts:115` — `if (envelope === null \|\| typeof envelope !== "object") return;` guard before push |
| W3 — ConfigWatcher gap undocumented | **FIXED** | `types.ts:47-60` — "Renderers are static across hot-reload" JSDoc with rationale + future `updateConfig` hook note |
| S1 — No test for render-after-stop | **FIXED** | `dashboard.test.ts:95-107` — "render() after stop() re-fills the buffer" test with intentionality documented |
| S2 — Shared makeEnvelope helper | Deferred | Agreed — test helpers naturally diverge |
| N1 — CLAUDE.md tree | Deferred | Agreed — low priority during migration series |
| N2 — Shutdown drain name near-collision | Deferred | Agreed — self-resolves at MIG-7.13 |

All cycle 1 blockers and warnings are resolved. Deferred items remain genuinely low-priority.

### Lens Results (Cycle 2 — fresh pass)

| Lens | Critical | Warning | Suggestion | Nit | Verdict |
|------|----------|---------|------------|-----|---------|
| CodeQuality | 0 | 0 | 0 | 1 | pass |
| Security | 0 | 0 | 0 | 0 | pass |
| Hardening | 0 | 0 | 0 | 0 | pass (N/A) |
| Architecture | 0 | 0 | 0 | 0 | pass |
| EcosystemCompliance | 0 | 0 | 0 | 0 | pass |
| Performance | 0 | 0 | 0 | 0 | pass |
| Duplication | 0 | 0 | 0 | 0 | pass |

### New Finding (1)

**N1 (nit/quality) — PagerDutyRenderer catch-block throw risk**
`pagerduty.ts:122` — catch block uses bare `envelope.id` without optional chaining. If a null envelope bypasses the surface-router and reaches `render()`, `buildPayload` throws → catch accesses `envelope.id` → TypeError inside catch → unhandled rejection. DashboardRenderer's catch uses `envelope?.id ?? "<null>"` (safe). Fix is one character: `envelope?.id ?? "<unknown>"`. Posted inline.

### What I verified and found clean

- **Security:** No injection surfaces, routingKey stays in POST body (never logged), endpoint override is constructor-only (not config-reachable), severity override validated against explicit allowlist. PD events-v2 auth model used correctly.
- **Architecture:** Clean module separation (types → dashboard/pagerduty → factory → cortex.ts). Dependencies flow one direction. `Renderer` → `RendererConfig` rename has zero in-tree consumers besides this PR. Two-stage validation (z.unknown → RendererSchema.safeParse) is the right additive migration pattern. surfaceConfig getter, shutdown drain, startup loop all mirror existing adapter code.
- **Performance:** Buffer bounded at 1000 with shift eviction. AbortSignal.timeout(5s) + AbortSignal.any composition on PD fetch. getRecent() snapshot copy is intentional. No growing caches, no blocking I/O.
- **Duplication:** surfaceConfig getter pattern matches existing adapter convention (Discord, Mattermost, Mock). PD abort signal composition vs common/timeout.ts fetchWithTimeout serve different purposes (signal composition vs diagnostic error rewriting). No re-implemented utilities.
- **Test quality:** 26 tests across 3 files cover happy path, failure paths, boundary conditions (buffer overflow, summary truncation), snapshot isolation, malformed input tolerance, post-stop behavior, custom endpoint, idempotent start.

**Review by jcfischer — COMMENTED** [review]
Full review (6 lenses + duplication) — cycle 2 confirms all W1+W2+W3+S1 fixes. One new nit: PD catch-block optional chaining asymmetry (inline). Clean across security, architecture, performance, duplication. 26/26 tests pin the contracts.

verdict: blockers=0 majors=0 nits=1 — recommend: merge (posted as comment-review — bot account opened the PR; --approve blocked by GitHub)
